frontend/hub/execution-environments/ExecutionEnvironmentForm.tsx

```diff
@@ -0,0 +1,333 @@
+import { useTranslation, TFunction } from 'react-i18next';
+import { useParams } from 'react-router-dom';
+import {
+  PageFormSubmitHandler,
+  PageFormTextInput,
+  PageHeader,
+  PageLayout,
+  useGetPageUrl,
+  PageFormTextArea,
+} from '../../../framework';
+import { TagIcon } from '@patternfly/react-icons';
+import { Button, InputGroup, Label, LabelGroup, TextInput } from '@patternfly/react-core';
+import { PageFormGroup } from '../../../framework/PageForm/Inputs/PageFormGroup';
+import { useGet } from '../../common/crud/useGet';
+import { HubRoute } from '../HubRoutes';
+import { hubAPI, hubAPIPost, pulpAPI } from '../api/utils';
+import { ExecutionEnvironment } from './ExecutionEnvironment';
+import { HubPageForm } from '../HubPageForm';
+import { HubItemsResponse } from '../useHubView';
+import { useState, useCallback, useMemo } from 'react';
+
+import { patchHubRequest, putHubRequest } from './../api/request';
+import { PageFormAsyncSelect } from '../../../framework/PageForm/Inputs/PageFormAsyncSelect';
+import { useSelectRegistrySingle } from './hooks/useRegistrySelector';
+import { usePageNavigate } from '../../../framework';
+
+export function CreateExecutionEnvironment() {
+  return <ExecutionEnvironmentForm mode="add" />;
+}
+
+export function EditExecutionEnvironment() {
+  return <ExecutionEnvironmentForm mode="edit" />;
+}
+
+function ExecutionEnvironmentForm(props: { mode: 'add' | 'edit' }) {
+  const page_size = 50;
+  const { t } = useTranslation();
+  const navigate = usePageNavigate();
+  const getPageUrl = useGetPageUrl();
+  const mode = props.mode;
+  const params = useParams<{ id?: string }>();
+
+  const [tagsToInclude, setTagsToInclude] = useState<string[]>([]);
+  const [tagsToExclude, setTagsToExclude] = useState<string[]>([]);
+  const [tagsSet, setTagsSet] = useState<boolean>(false);
+
+  const registry = useGet<HubItemsResponse<Registry>>(
+    hubAPI`/_ui/v1/execution-environments/registries?page_size=${page_size.toString()}`
+  );
+
+  const eeUrl = useMemo(
+    () =>
+      mode == 'edit'
+        ? hubAPI`/v3/plugin/execution-environments/repositories/${params?.id || ''}/`
+        : '',
```

Comment:
`params?.id || ''` implies params.id can be an empty string...

such an url would return a list of repositories in dev mode, and a 404 in production => not what we want :)

```suggestion
      mode == 'edit' && params?.id
        ? hubAPI`/v3/plugin/execution-environments/repositories/${params.id}/`
        : '',
```

Reply:
(also, strings are compared by value, not object identity, so the `useMemo` is probably superfluous, but up to you :))

Reply:
You are right, it works. I was not sure how it is compared in SWR (if it was identity, it would need useMemo). I was there a little paranoid.